ccpp/data/GFS_typedefs.F90

```diff
@@ -2157,7 +2160,10 @@ module GFS_typedefs
     real (kind=kind_phys), pointer      :: qs_lay(:,:)               => null()  !<
     real (kind=kind_phys), pointer      :: q_lay(:,:)                => null()  !<
     real (kind=kind_phys), pointer      :: deltaZ(:,:)               => null()  !<
+    real (kind=kind_phys), pointer      :: deltaZc(:,:)              => null()  !< 
+    real (kind=kind_phys), pointer      :: deltaP(:,:)               => null()  !< 
     real (kind=kind_phys), pointer      :: cloud_overlap_param(:,:)  => null()  !< Cloud overlap parameter
+    real (kind=kind_phys), pointer      :: cnv_cloud_overlap_param(:,:) => null()  !< Convective cloud overlap parameter
     real (kind=kind_phys), pointer      :: precip_overlap_param(:,:) => null()  !< Precipitation overlap parameter
     real (kind=kind_phys), pointer      :: tracer(:,:,:)             => null()  !<
```

Comment:
Is it possible to use tracer array qgrs instead of declaring this new array? Because this array could be huge, it needs to be careful to declare such a large array.

Reply:
The mixing-ratios need to be sanitized before passing them to the radiation. 
Here in RRTMGP, tracer is an interstitial. In RRTMG, tracer is a local field in GFS_rrtmg_pre.F90.

Reply:
But I did't see differences between variable tracer and qgrs in GFS_rrtmgp_pre.F90 except tracers (except sphum) are set to > 0 in variable tracer.  And the dimensions between tracer and qgrs are exactly same. See below:

    do j = 2, nTracers
       tracer(1:NCOL,:,j) = qgrs(1:NCOL,:,j)
       where(tracer(:,:,j) .lt. 0.0) tracer(:,:,j) = 0._kind_phys
    enddo

Moreover, those GFS_rrtmgp_microphysics-scheme_pre.F90 files used index name of tracers, so it doens't matter whether tracer includes sphum or not.  Those rrtmgp_?w_aerosol_optics.F90 files call subroutine setaer, which is inside radiation_aerosols.f where tracer is not actively used, even if it used there, tracer can be easily deal with.

Reply:
@ChunxiZhang-NOAA 
I agree that the tracer/qgrs can be harmonized. I will make some changes later this afternoon.

Reply:
You probably don't need tsfg.  p_lay, t_lay, p_lev, t_lev, and q_lay are recalculated in GFS_rrtmgp_pre.F90. Mostly just set boundaries and interpolations. Are those boundaries and interpolations necessary?

Reply:
Yeah. GP is picky and needs the inputs to be in range of its LUTs.

Reply:
I see, thanks.